### PR TITLE
Different error codes for missing D2L files based on role

### PR DESCRIPTION
**Pull request by marcospri**
For:

- https://github.com/hypothesis/lms/issues/4946

#### Instructor:

- Login as `HypothesisEng.Teacher`
- Launch https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/2376/View?ou=6782

#### Learner

- Login as `aunltd.S1`
- Launch https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/2376/View?ou=6782

### CODE REVIEW

File: lms/services/d2l_api/client.py [hunk 1]
```diff
@@ -64,10 +64,10 @@ class D2LTableOfContentsSchema(RequestsResponseSchema):
 
 
 class D2LAPIClient:
-    def __init__(self, basic_client, request, file_service):
+    def __init__(self, basic_client, file_service, lti_user):
```
Comment (marcospri) [inline lms/services/d2l_api/client.py:67]: request was not being used :shrug: 

Added LTIUser here.

File: lms/views/api/d2l/files.py [hunk 2]
```diff
@@ -46,7 +46,9 @@ def via_url(_context, request):
     try:
         if request.lti_user.is_instructor:
             if not course_copy_plugin.is_file_in_course(course_id, file_id):
-                raise FileNotFoundInCourse("d2l_file_not_found_in_course", file_id)
+                raise FileNotFoundInCourse(
+                    "d2l_file_not_found_in_course_instructor", file_id
```
Comment (marcospri) [inline lms/views/api/d2l/files.py:50]: Always instructor in this branch.

File: lms/static/scripts/ui-playground/components/ErrorComponents.tsx [hunk 3]
```diff
@@ -466,7 +466,8 @@ export default function ErrorComponents() {
             <div className="grid grid-cols-2 gap-2">
               {(
                 [
-                  'd2l_file_not_found_in_course',
+                  'd2l_file_not_found_in_course_student',
```
Comment (marcospri) [inline lms/static/scripts/ui-playground/components/ErrorComponents.tsx:469]: @lyzadanger Is this all it's needed to fix them for the ui-playground? 

I don't know how to check that locally.

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx [hunk 4]
```diff
@@ -117,15 +117,45 @@ export default function LaunchErrorDialog({
         </ErrorModal>
       );
 
-    case 'd2l_file_not_found_in_course':
+    case 'd2l_file_not_found_in_course_instructor':
```
Comment (marcospri) [inline lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx:120]: Different code for different role. If we start using this pattern more we might want to look into adding more metadata to the error response (role, product...)

For now just a different error code is simple enought.

File: lms/static/scripts/ui-playground/components/ErrorComponents.tsx (hunk 3 above)
Comment (lyzadanger) [reply lms/static/scripts/ui-playground/components/ErrorComponents.tsx:469]: It's easy! Go here: http://localhost:8001/ui-playground/errors and find the error you want to look at: 

image

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx [hunk 5]
```diff
@@ -117,15 +117,45 @@ export default function LaunchErrorDialog({
...
       return (
         <ErrorModal
           {...defaultProps}
           title="Hypothesis couldn't find the file in the course"
         >
+          <p>This might have happened because:</p>
+          <ul className="px-4 list-disc">
+            <li>The file has been deleted from D2L</li>
+            <li>
+              The course was copied and the selected is not in the new course.
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx:130]:
```suggestion
              The course was copied and the selected file is not available in the new course.
```

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx [hunk 6]
```diff
@@ -117,15 +117,45 @@ export default function LaunchErrorDialog({
...
+            </li>
+          </ul>
+          <p>
+            To fix the issue recreate the assignment with a different file. Also
+            check our document about
+            <ExternalLink href="https://web.hypothes.is/help/using-hypothesis-with-d2l-course-content-files/">
+              Using Hypothesis With D2L Course Content Files
+            </ExternalLink>
+          </p>
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx:139]: What does _recreate_ mean here, exactly? Is the implication that (without edit-assignment support) there is no way to fix the problem without creating a new assignment?

Note the explicit `{' '}` to ensure that there is a space between _about_ and the link. 

```suggestion
          <p>
            To fix the issue, recreate the assignment with a different file. More information can be found in our document about{' '}
            <ExternalLink href="https://web.hypothes.is/help/using-hypothesis-with-d2l-course-content-files/">
              Using Hypothesis With D2L Course Content Files
            </ExternalLink>.
          </p>
```

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx [hunk 7]
```diff
@@ -117,15 +117,45 @@ export default function LaunchErrorDialog({
...
+      return (
+        <ErrorModal
+          {...defaultProps}
+          title="Hypothesis couldn't find the file in the course"
+        >
+          <p>This might have happened because:</p>
+          <ul className="px-4 list-disc">
+            <li>The file has been deleted from D2L</li>
+            <li>
+              The course was copied and the selected is not in the new course.
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx:153]:
```suggestion
              The course was copied and the selected file is not available in the new course.
```

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx [hunk 8]
```diff
@@ -117,15 +117,45 @@ export default function LaunchErrorDialog({
...
+            </li>
+          </ul>
           <p>
-            This might have happened because the file has been deleted from D2L.
-            An instructor needs to re-create the assignment with a new file.
+            Please contact the course instructor pointing them to this
+            assignment.
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx:158]: Maybe this? I dunno.

```suggestion
            Please ask the course instructor to review the settings of this assignment.
```

**Review by lyzadanger — CHANGES_REQUESTED** [review]
(Front-end pieces only). Hi @marcospri! I added a few suggestions for wording of the errors.

File: lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx (hunk 6 above)
Comment (marcospri) [reply lms/static/scripts/frontend_apps/components/LaunchErrorDialog.tsx]: Yeah, there's no way out of this other than to recreate the assignment in the LMS, the same is true for other LMS for both files and group issues.

The same wording/solution is issued in KB articles etc. Editing assignments will be a big deal for this issues.

File: lms/services/d2l_api/client.py [hunk 9]
```diff
@@ -152,7 +152,10 @@ def public_url(self, org_unit, file_id) -> str:
         except ExternalRequestError as err:
             if err.status_code == 404:
                 raise FileNotFoundInCourse(
-                    "d2l_file_not_found_in_course", file_id
+                    "d2l_file_not_found_in_course_instructor"
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply lms/services/d2l_api/client.py:155]: For sure. I've fallen into the "hey we can construct this error code, that'll be nice" trap before. Definitely better to be direct about it.

File: tests/unit/lms/services/d2l_api/client_test.py [hunk 10]
```diff
@@ -210,15 +210,34 @@ def test_public_url(self, svc, basic_client):
...
+        self,
+        svc,
+        basic_client,
+        status_code,
+        user_fixture_name,
+        exception,
+        error_code,
+        request,
     ):
+        _ = request.getfixturevalue(user_fixture_name)
```
Comment (jon-betts) [inline tests/unit/lms/services/d2l_api/client_test.py:240]: Interesting! Why do we assign to `_` here?

**Review by jon-betts — APPROVED** [review]
The Python part of this is pretty simple. I don't have any feedback really. 

The use of `request` in the tests is interesting. I wonder if we are missing some tricks here in general?

**Pull request merged**